src/main/java/org/spongepowered/common/event/tracking/phase/block/DispensePhaseState.java

```diff
@@ -58,8 +69,44 @@ public void unwind(final GeneralizedContext phaseContext) {
             });
         phaseContext.getCapturedEntitySupplier()
             .acceptAndClearIfNotEmpty(entities -> {
-                Sponge.getCauseStackManager().addContext(EventContextKeys.SPAWN_TYPE, SpawnTypes.DISPENSE);
-                SpongeCommonEventFactory.callSpawnEntity(entities, phaseContext);
+                List<Entity> projectiles = new ArrayList<>();
+                List<Entity> rest = new ArrayList<>();
+                for (Entity entity : entities) {
+                    if (entity instanceof Projectile) {
+                        projectiles.add(entity);
+                    } else {
+                        rest.add(entity);
+                    }
+                }
+
+                if (!rest.isEmpty()) {
+                    Sponge.getCauseStackManager().addContext(EventContextKeys.SPAWN_TYPE, SpawnTypes.DISPENSE);
```

Comment:
Should not be adding to the context outside a frame, or without popping the context afterwards.